driver:adc: enable the ADC threshold configuration and wake-up functi…

JacobLi29 opened the pull request:
…on support.

Commit:
driver:adc: enable the ADC threshold configuration and wake-up function support.

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -189,14 +189,16 @@ static void adc_npcx_isr(const struct device *dev)
+
 	/* Clear status pending bits first */
 	inst->ADCSTS = status;
 	LOG_DBG("%s: status is %04X\n", __func__, status);
```
@JacobLi29, At Line 250, despite enabling CONFIG_ADC_CMP_NPCX in prj.conf, the macro does not appear to be enabled in the .config file within the build folder.

It has been observed that ADC_CMP_NPCX does not have a configuration member in Kconfig.npcx. 

Could you please check on this

JacobLi29 replied:
ADC_CMP_NPCX is defined in `zephyr\drivers\sensor\nuvoton\nuvoton_adc_cmp_npcx\Kconfig`. It enables the ADC threshold function configured by the DTS file. I will remove the CONFIG_ADC_CMP_NPCX condition to enable threshold status check directly.

anandkumarns replied:
_" I will remove the CONFIG_ADC_CMP_NPCX condition to enable threshold status check directly."_

Sure @JacobLi29

JamesChang1017 replied:
"remove the CONFIG_ADC_CMP_NPCX condition to enable threshold status check directly"
Hi Jacob, can you please update by today?

anandkumarns replied:
@JacobLi29 , Could you please mainline the change. This is will be required as part of EC FW v1.14 release

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -748,6 +746,138 @@ int adc_npcx_threshold_mv_to_thrval(const struct device *dev, uint32_t val_mv,
+
+	/* Enable/Disable threshold IRQ */
+	if (enable) {
+		/* Set control threshold registers */
```
Better to turn off the ADC , during the register configuration. that helps to avoid unwanted blocking of Sleep
		/* Turn off ADC */
		inst->ADCCNF &= ~BIT(NPCX_ADCCNF_ADCEN);

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -748,6 +746,138 @@ int adc_npcx_threshold_mv_to_thrval(const struct device *dev, uint32_t val_mv,
+		thrcts |= BIT(NPCX_THRCTS_THR1_IEN + th_sel);
+
+		inst->THRCTS = thrcts;
+
```
We can perform ADC channel Data, to avoid causing the overflow switching from thermal management adc read (Single shot) to group of channel sampling
ret_thrcts = CHNDAT(config->base, t_ctrl->chnsel);

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -748,6 +746,138 @@ int adc_npcx_threshold_mv_to_thrval(const struct device *dev, uint32_t val_mv,
+		t_data->repetitive_channels |= BIT(t_ctrl->chnsel);
+		adc_npcx_config_channels(dev, t_data->repetitive_channels);
+	} else {
+		/* avoid clearing other threshold status */
```
Clearing the status register prevents the ISR function from being called post exit from sleep. 

Could you please review and consider commenting out or removing this line?

JacobLi29 replied:
The status should be cleared while disabling the threshold interrupt and the threshold detection. If it is not cleared, it will not be serviced while the threshold interrupt is disabled.

anandkumarns replied:
Sure Got it. Post changing the below api to notify_pm_state_exit.    
/* disable the ADC threshold wake-up interrupt */
    adc_thrm_pwrdn_exit();
    
    1. Exit from sleep
    2. Enable the interrupt 
    3. ISR will get called if ADC comparator low/high crosses
    4. notify_pm_state_exit (clear the adc wakeup and clear the threshold Interrupt)

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -748,6 +746,138 @@ int adc_npcx_threshold_mv_to_thrval(const struct device *dev, uint32_t val_mv,
+	uint8_t status;
+
+	if(enable) {
+		/* Turn on ADC */
```
Could you please comment/delete the ADC ON here rather we can enable after the ADC configuration. 
at line 862
		/* Turn on ADC */
		inst->ADCCNF |= BIT(NPCX_ADCCNF_ADCEN);

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -748,6 +746,138 @@ int adc_npcx_threshold_mv_to_thrval(const struct device *dev, uint32_t val_mv,
+
+		/* Set conversion type to repetitive (runs continuously) */
+		inst->ADCCNF |= BIT(NPCX_ADCCNF_ADCRPTC);
+
```
we can add here, 
/* Turn on ADC */
inst->ADCCNF |= BIT(NPCX_ADCCNF_ADCEN);

anandkumarns commented on drivers/adc/adc_npcx.c:
```diff
@@ -748,6 +746,138 @@ int adc_npcx_threshold_mv_to_thrval(const struct device *dev, uint32_t val_mv,
+		/* No thesholdd active left, disable repetitive mode */
+		inst->ADCCNF &= ~BIT(NPCX_ADCCNF_ADCRPTC);
+		/* Turn off ADC */
+		inst->ADCCNF &= ~BIT(NPCX_ADCCNF_ADCEN);
```
Disabling the ADC will require explicit re-enabling to resume its functionality?

Observations indicate that adc_read is encountering issues, as the managing thread remains inactive even after resumption.

JacobLi29 replied:
the ADC function will be re-enabled while the managing thread resumes and calls the adc_read() function, or configures the threshold detection during transaction into deep sleep mode.

anandkumarns replied:
Got it. when i was checking with EC FW v1.11 or v1.12. post exit from sleep with ADC off. thermal and Fan management threads were not waking up properly. 

With EC FW v1.13, Even turning off the ADC. both FAN and thermal management thread works fine.

Commit:
driver:adc: remove the CONFIG_ADC_CMP_NPCX condition to enable threshold status check.
#1 The ADC threshold event check and service in ISR routine will be blocked while CONFIG_ADC_CMP_NPCX is not enabled. Remove this condition to let threshold event will always be serviced.
